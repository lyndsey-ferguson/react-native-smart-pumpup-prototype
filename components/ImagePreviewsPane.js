/**
 ImagePreviewsPane.js
 Smart Components

 React Native app created by Lyndsey on 10/7/17.
 Copyright Â© 2017 Lyndsey Ferguson Apps. All rights reserved.
*/

import React, {
  Component
} from 'react';
import {
  Dimensions,
  FlatList,
  Image,
  StyleSheet,
  View
} from 'react-native';
import Bounceable from 'react-native-bounceable';

const { width: windowWidth } = Dimensions.get('window');
const imagePreviewsPaneWidth = windowWidth - 40;

export default class ImagePreviewsPane extends React.Component {
  renderImagePreview({item}) {
    return (
      <Image style={styles.imagePreview}
             resizeMode='contain'
             source={{uri: item.uri}}
             key={item.key}/>
    );
  }
  handleScrollEnd(event) {
    const horizontalOffset = event.nativeEvent.contentOffset.x;
    const imagePreviewIndex = horizontalOffset ? Math.round(horizontalOffset / imagePreviewsPaneWidth) : 0;
    if (this.props.currentImageIndex != imagePreviewIndex) {
      this.props.ChangeCurrent(imagePreviewIndex);
    }
  }
  getItemLayout(data, index) {
    return {
      length: imagePreviewsPaneWidth,
      offset: imagePreviewsPaneWidth * index,
      index
    };
  }

  render() {
    const { currentImageIndex, imagesList } = this.props;
    const shouldScroll = currentImageIndex < imagesList.length && currentImageIndex > 0;
    return (
      <View>
        <FlatList ref={component => {
                    this.flatList = component;
                    if (component && shouldScroll) {
                      component.scrollToIndex({index: currentImageIndex});
                    }}
                  }
                  onMomentumScrollEnd={(event) => this.handleScrollEnd.call(this, event)}
                  extraData={imagesList}
                  data={imagesList}
                  style={styles.imagePreviews}
                  horizontal={true}
                  decelerationRate={0}
                  getItemLayout={this.getItemLayout.bind(this)}
                  snapToAlignment={'center'}
                  snapToInterval={imagePreviewsPaneWidth}
                  renderItem={this.renderImagePreview.bind(this)} />
        <View
          style={{ flexDirection: 'row', justifyContent: 'center' }} // this will layout our dots horizontally (row) instead of vertically (column)
          >
          {
            imagesList.map((_, imageIndex) => {
              return (
                <Bounceable key={imageIndex}
                            onPress={() => this.props.ChangeCurrent(imageIndex)}
                            level={1.1}
                >
                  <View style={[
                    styles.imagePreviewDot,
                    currentImageIndex == imageIndex && styles.imagePreviewActiveDot
                  ]} />
                </Bounceable>
              );
            })}
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  imagePreviews: {
    width: imagePreviewsPaneWidth,
    height: imagePreviewsPaneWidth,
    marginTop: 20,
    marginLeft: 20
  },
  imagePreview: {
    width: imagePreviewsPaneWidth,
    height: imagePreviewsPaneWidth,
  },
  imagePreviewDot: {
    height: 10,
    width: 10,
    backgroundColor: '#595959',
    margin: 8,
    borderRadius: 5
  },
  imagePreviewActiveDot: {
    backgroundColor: '#FFFF00',
  }
});
