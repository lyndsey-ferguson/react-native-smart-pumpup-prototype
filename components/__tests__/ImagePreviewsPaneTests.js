/**
 ImagePreviewsPaneTests.js
 Smart Components

 React Native app created by Lyndsey on 10/7/17.
 Copyright Â© 2017 Lyndsey Ferguson Apps. All rights reserved.
*/
import React from 'react'
import ImagePreviewsPane from '../ImagePreviewsPane'
import renderer from 'react-test-renderer'
import expect from 'expect'



it('renders correctly without any images and a currentImageIndex of 0', () => {
  const tree = renderer.create(
    <ImagePreviewsPane imagesList={[]} currentImageIndex={9}/>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})



it('renders correctly without any images and an invalid currentImageIndex', () => {
  const tree = renderer.create(
    <ImagePreviewsPane imagesList={[]} currentImageIndex={9}/>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})



it('renders correctly with 6 images and a currentImageIndex of 0', () => {
  const tree = renderer.create(
    <ImagePreviewsPane  currentImageIndex={0}
      imagesList={[
        { uri: 'https://www.example.com/image1.png', key: '1'},
        { uri: 'https://www.example.com/image2.png', key: '2'},
        { uri: 'https://www.example.com/image3.png', key: '3'},
        { uri: 'https://www.example.com/image4.png', key: '4'},
        { uri: 'https://www.example.com/image5.png', key: '5'},
        { uri: 'https://www.example.com/image6.png', key: '6'},
      ]}/>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
