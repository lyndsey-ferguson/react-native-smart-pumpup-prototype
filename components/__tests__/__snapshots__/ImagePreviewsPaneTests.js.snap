// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with 6 images and a currentImageIndex of 0 1`] = `
<View>
  <RCTScrollView
    data={
      Array [
        Object {
          "key": "1",
          "uri": "https://www.example.com/image1.png",
        },
        Object {
          "key": "2",
          "uri": "https://www.example.com/image2.png",
        },
        Object {
          "key": "3",
          "uri": "https://www.example.com/image3.png",
        },
        Object {
          "key": "4",
          "uri": "https://www.example.com/image4.png",
        },
        Object {
          "key": "5",
          "uri": "https://www.example.com/image5.png",
        },
        Object {
          "key": "6",
          "uri": "https://www.example.com/image6.png",
        },
      ]
    }
    decelerationRate={0}
    disableVirtualization={false}
    extraData={
      Array [
        Object {
          "key": "1",
          "uri": "https://www.example.com/image1.png",
        },
        Object {
          "key": "2",
          "uri": "https://www.example.com/image2.png",
        },
        Object {
          "key": "3",
          "uri": "https://www.example.com/image3.png",
        },
        Object {
          "key": "4",
          "uri": "https://www.example.com/image4.png",
        },
        Object {
          "key": "5",
          "uri": "https://www.example.com/image5.png",
        },
        Object {
          "key": "6",
          "uri": "https://www.example.com/image6.png",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    horizontal={true}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={undefined}
    renderItem={[Function]}
    scrollEventThrottle={50}
    snapToAlignment="center"
    snapToInterval={710}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "height": 710,
        "marginLeft": 20,
        "marginTop": 20,
        "width": 710,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View>
      <View
        onLayout={undefined}
        style={null}
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://www.example.com/image1.png",
            }
          }
          style={
            Object {
              "height": 710,
              "width": 710,
            }
          }
        />
      </View>
      <View
        onLayout={undefined}
        style={null}
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://www.example.com/image2.png",
            }
          }
          style={
            Object {
              "height": 710,
              "width": 710,
            }
          }
        />
      </View>
      <View
        onLayout={undefined}
        style={null}
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://www.example.com/image3.png",
            }
          }
          style={
            Object {
              "height": 710,
              "width": 710,
            }
          }
        />
      </View>
      <View
        onLayout={undefined}
        style={null}
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://www.example.com/image4.png",
            }
          }
          style={
            Object {
              "height": 710,
              "width": 710,
            }
          }
        />
      </View>
      <View
        onLayout={undefined}
        style={null}
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://www.example.com/image5.png",
            }
          }
          style={
            Object {
              "height": 710,
              "width": 710,
            }
          }
        />
      </View>
      <View
        onLayout={undefined}
        style={null}
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": "https://www.example.com/image6.png",
            }
          }
          style={
            Object {
              "height": 710,
              "width": 710,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      collapsable={undefined}
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#595959",
                  "borderRadius": 5,
                  "height": 10,
                  "margin": 8,
                  "width": 10,
                },
                Object {
                  "backgroundColor": "#FFFF00",
                },
              ]
            }
          />
        </View>
      </View>
    </View>
    <View
      collapsable={undefined}
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#595959",
                  "borderRadius": 5,
                  "height": 10,
                  "margin": 8,
                  "width": 10,
                },
                false,
              ]
            }
          />
        </View>
      </View>
    </View>
    <View
      collapsable={undefined}
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#595959",
                  "borderRadius": 5,
                  "height": 10,
                  "margin": 8,
                  "width": 10,
                },
                false,
              ]
            }
          />
        </View>
      </View>
    </View>
    <View
      collapsable={undefined}
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#595959",
                  "borderRadius": 5,
                  "height": 10,
                  "margin": 8,
                  "width": 10,
                },
                false,
              ]
            }
          />
        </View>
      </View>
    </View>
    <View
      collapsable={undefined}
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#595959",
                  "borderRadius": 5,
                  "height": 10,
                  "margin": 8,
                  "width": 10,
                },
                false,
              ]
            }
          />
        </View>
      </View>
    </View>
    <View
      collapsable={undefined}
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#595959",
                  "borderRadius": 5,
                  "height": 10,
                  "margin": 8,
                  "width": 10,
                },
                false,
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly without any images and a currentImageIndex of 0 1`] = `
<View>
  <RCTScrollView
    data={Array []}
    decelerationRate={0}
    disableVirtualization={false}
    extraData={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    horizontal={true}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={undefined}
    renderItem={[Function]}
    scrollEventThrottle={50}
    snapToAlignment="center"
    snapToInterval={710}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "height": 710,
        "marginLeft": 20,
        "marginTop": 20,
        "width": 710,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  />
</View>
`;

exports[`renders correctly without any images and an invalid currentImageIndex 1`] = `
<View>
  <RCTScrollView
    data={Array []}
    decelerationRate={0}
    disableVirtualization={false}
    extraData={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    horizontal={true}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={undefined}
    renderItem={[Function]}
    scrollEventThrottle={50}
    snapToAlignment="center"
    snapToInterval={710}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "height": 710,
        "marginLeft": 20,
        "marginTop": 20,
        "width": 710,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  />
</View>
`;
